unified-scheduler-pool/src/lib.rs

```diff
@@ -558,61 +578,78 @@ impl<S: SpawnableScheduler<TH>, TH: TaskHandler> ThreadManager<S, TH> {
-                                        .unwrap();
-                                    active_task_count = active_task_count.checked_add(1).unwrap();
-                                }
-                                NewTaskPayload::OpenSubchannel(context) => {
-                                    // signal about new SchedulingContext to handler threads
-                                    runnable_task_sender
-                                        .send_chained_channel(context, handler_count)
-                                        .unwrap();
-                                    assert_matches!(
-                                        result_with_timings.replace(initialized_result_with_timings()),
-                                        None
-                                    );
-                                }
-                                NewTaskPayload::CloseSubchannel => {
-                                    session_ending = true;
-                                }
-                            }
-                        },
-                    };
+            move || {
+                let (do_now, dont_now) = (&disconnected::<()>(), &never::<()>());
+                let dummy_receiver = |trigger| {
+                    if trigger {
+                        do_now
+                    } else {
+                        dont_now
+                    }
+                };
 
-                    // a really simplistic termination condition, which only works under the
-                    // assumption of single handler thread...
-                    is_finished = session_ending && active_task_count == 0;
-                }
+                let mut state_machine = unsafe {
+                    SchedulingStateMachine::exclusively_initialize_current_thread_for_scheduling()
+                };
 
-                if session_ending {
-                    session_result_sender
-                        .send(Some(
-                            result_with_timings
-                                .take()
-                                .unwrap_or_else(initialized_result_with_timings),
-                        ))
-                        .unwrap();
-                    session_ending = false;
+                loop {
+                    let mut is_finished = false;
+                    while !is_finished {
+                        select! {
+                            recv(finished_task_receiver) -> executed_task => {
+                                let executed_task = executed_task.unwrap();
+
+                                state_machine.deschedule_task(&executed_task.task);
+                                let result_with_timings = result_with_timings.as_mut().unwrap();
+                                Self::accumulate_result_with_timings(result_with_timings, executed_task);
+                            },
+                            recv(dummy_receiver(state_machine.has_unblocked_task())) -> dummy => {
```

Comment:
> The dummy receiver pattern seems really odd to me, though I see how it is working
>
> .........
>
> Anyway, my initial thought on the oddity of the pattern still stands. So not going to delete my comment, and still request your thoughts.

Hope I explained the justification clearly: 89b7773288187966419b9861d5d4fb0bd5981413. I think you can get used to it. this pattern is simplest and no redundant code and most performant. :)

> I'd actually much prefer we make this very clear since I don't think it's a very obvious thing:
> 
> ```rust
> let dummy_receiver = dummy_receiver(state_machine.has_unblocked_task());
> select!{
>     // ...
> }
> ```

This is done with comment. As I ranted in the comment, ideally, I want to move the `dummy_receiver()` invocation back into the `select!`...

> I know you had made a branch for a biased select - is the order here i.e. `finished -> dummy -> new` the biased order you want?

Yes, this is correct.



> It could be overkill for the current MNB load we have, but I wonder if you considered just running a hot loop here if we are actively scheduling; not sure if using try_recvs in hot loop would cause more contention on the channels, you're certainly more familiar with crossbeam internals.

As I explicit commented on in the commit 89b7773288187966419b9861d5d4fb0bd5981413. it's too early to take the busy loop path and i haven't fully grokked the perf implication of busy looping. Casual tests showed an improved latency indeed, though.

> But basically I am asking about something along the lines of:
> 
> ```
> ...
> ```

By the way, busy looping can be done with `select{,_biased}!` pretty easily like this:

```rust
loop ...
    select{_biased}! {
        recv(...) => ....,
        ...,
        default => continue, // there's special-cased the `default` selector in `select{,_biased}!`
    }
    ...
    if is_finished {
        break;
    }

}
```